ASoC: SOF: Intel: hda-mlink: add support for Extended Multi Links

plbossart opened the pull request:
The LunarLake hardware moved a set of DMIC/SSP/SoundWire registers in the HDaudio multi-link areas. This PR adds a set of helpers to configure the DMIC/SSP/SoundWire. In the case of SoundWire, the configurations will be made both from the sound/soc/sof/intel and drivers/soundwire/intel locations, so the hda-mlink is now a separate module to avoid circular dependencies.

The actual use of these helpers will be provided when this PR https://github.com/thesofproject/linux/pull/4158 and PR #4157 are merged. There will be cross-dependencies between ALSA, ASoC and SoundWire so we have to progress in steps.

One possible pushback is that this adds yet another layer on top of hdac_link and hdac_ext_link. That's a valid point. The reason why this was added as a separate extension is that we don't have the ability to validate the Skylake and AVS drivers, and there's no reason why these drivers would need to know about a hardware extension added 8 years after the introduction of SKL/KBL.

One could also argue that this extension is not SOF-specific. That's correct from a hardware perspective, however since there are no other users it's not clear where else than sound/soc/sof/intel this module should be located.

@tiwai FYI

plbossart commented:
2 checkpatch warnings missed, oops. fixed now.

plbossart commented:
@ujfalusi @RanderWang @ranj063 @bardliao can you please look into this PR? it's rather important and gates the rest of the SoundWire patches.

RanderWang commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -14,28 +14,235 @@
+				return ret;
+			}
+		}
+	}
```
how about 
````c
if (!bus->mlcap)
      return 0;

link_count = readl(bus->mlcap + AZX_REG_ML_MLCD) + 1;
...
````

plbossart replied:
yes, thanks for the suggestion. I modified the previous code using the same test, but it's better to invert the logic.

plbossart replied:
done

RanderWang commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -324,6 +324,49 @@ int hdac_bus_eml_get_count(struct hdac_bus *bus, bool alt, int elid)
+	else
+		lctl &= ~AZX_ML_LCTL_INTEN;
+	writel(lctl, hlink->ml_addr + AZX_REG_ML_LCTL);
+
```
snd_hdac_updatel(hlink->ml_addr, AZX_REG_ML_LCTL, AZX_ML_LCTL_INTEN, enable);

RanderWang commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -761,6 +773,30 @@ struct hdac_ext_link *hdac_bus_eml_dmic_get_hlink(struct hdac_bus *bus)
+
+	mutex_lock(&h2link->eml_lock);
+
+	hdaml_lctl_offload_enable(hlink->ml_addr + AZX_REG_ML_LCTL, enable);
```
````c
val = enable ?  AZX_ML_LCTL_OFLEN : 0;
snd_hdac_updatel(hlink->ml_addr, AZX_REG_ML_LCTL, AZX_ML_LCTL_OFLEN, val);
````
I find a few bit operation functions in the same style.

plbossart replied:
no. I don't want any dependency on an external library. These routines follow the HAS as is.

bardliao commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -737,6 +737,30 @@ int hda_bus_ml_suspend(struct hdac_bus *bus)
+{
+	struct hdac_ext2_link *h2link;
+
+	h2link = find_ext2_link(bus, true, AZX_REG_ML_LEPTR_ID_INTEL_SSP);
```
Should `AZX_REG_ML_LEPTR_ID_INTEL_SSP` be `AZX_REG_ML_LEPTR_ID_INTEL_DMIC`?

plbossart replied:
yes indeed, good old copy-paste mistake. Thank you Sir!

plbossart replied:
fixed

ranj063 commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -14,28 +14,235 @@
+		hlink->lsdiid = readw(ml_addr + AZX_REG_ML_LSDIID);
+
+		dev_dbg(dev, "Link %d: HDAudio - lsdiid=%d\n",
+			link_idx, hlink->lsdiid);
```
why not return 0 here so you can reduce the indentation for the else part?

plbossart replied:
yes indeed.

plbossart replied:
done

ranj063 commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -14,28 +14,235 @@
+
+/* END HDAML section */
+
+static int hda_ml_alloc_h2link(struct hdac_bus *bus, int i)
```
can we call this second arg index as well, would be cleaner to read

plbossart replied:
yes

plbossart replied:
done

ranj063 commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -14,28 +14,235 @@
+		for (i = 0; i < link_count; i++) {
+			ret = hda_ml_alloc_h2link(bus, i);
+			if (ret < 0) {
+				hda_bus_ml_free(bus);
```
assuming h2link alloc failed with -ENOMEM, calling hda_bus_ml_free() would end up calling mutex_destroy(&h2link->eml_lock); but h2link would be NULL here isnt it?

plbossart replied:
I don't think so, we only add the link to the list when there's no error, so on free that link would be skipped.

In other words, the hda_bus_ml_free() is to free what was allocated in the previous iterations.

ranj063 commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -167,6 +167,56 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
+	u32 val;
+
+	udelay(3);
+	retry = 150;
```
can be initialized during declaration?

plbossart replied:
done

ranj063 commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -167,6 +167,56 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
+	int retry;
+	u32 val;
+
+	udelay(3);
```
do we need a macro for the 3 us?

plbossart replied:
I need to check what the recommended values are, I don't even recall where this 3 and 150 values come from.

plbossart replied:
done

ranj063 reviewed (commented):
looks great. just minor comments

plbossart commented:
v2: addressed comments, moved interrupt code in hdaml section.
Still to be done; revisit what the sleep and retry times should be with hardware team.

plbossart commented:
v3: modified hardware poll delay to 10us and max 1ms

ujfalusi commented on include/sound/hda-mlink.h:
```diff
@@ -53,6 +56,12 @@ int hdac_bus_eml_power_down_unlocked(struct hdac_bus *bus, bool alt, int elid, i
+int hdac_bus_eml_sdw_power_up_unlocked(struct hdac_bus *bus, int sublink) { return 0; }
+
+static inline
+int hdac_bus_eml_sdw_power_down_unlocked(struct hdac_bus *bus, int sublink) {return 0; }
```
The common way to break the line like this is to keep the 'static inline int` in one line:
```c
static inline int
hdac_bus_eml_sdw_power_down_unlocked(struct hdac_bus *bus, int sublink) { return 0; }
```

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -324,6 +324,18 @@ find_ext2_link(struct hdac_bus *bus, bool alt, int elid)
+	if (!h2link)
+		return 0;
+
+	return h2link->slcount;
```
In other places (if I recall right) have the opposite logic:
```c
if (h2link)
	return h2link->slcount;
return 0;
```

ujfalusi replied:
and also: for HDA (alt=0) slcount is 0

plbossart replied:
There's no test with if(h2link)

The pattern is always

h2link = find_ext2_link(bus, alt, elid);
if (!h2link)
	return error;
	
And yes for HDaudio the intent is that the count is zero. 

No change here, not sure if I got the feedback

plbossart replied:
actually no, for HDAudio the count is one and has been initialized now.

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -228,6 +228,26 @@ static int hdaml_link_shutdown(u32 __iomem *lctl, int sublink)
+{
+	u32 val;
+
+	val = readl(lctl);
```
can be in one line:
```c
u32 val = readl(lctl);
```

plbossart replied:
I try to avoid grouping those. I prefer a clear rmw block, that's really a matter of personal preference.

plbossart replied:
actually I did this for other functions, so I'll make the code consistent.

RanderWang reviewed (dismissed):
LGTM

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -248,6 +309,113 @@ void hda_bus_ml_free(struct hdac_bus *bus)
+		struct hdac_ext2_link *h2link = hdac_ext_link_to_ext2(hlink);
+
+		if (h2link->alt == alt && h2link->elid == elid)
+			return h2link;
```
h2link->elid is only initialized for extended links (non HDA will have elid == 0 ?)
I guess it is fine to expect 0 for HDA, just asking.

plbossart replied:
The combination is this:

ALT 0 elid 0 -> HDaudio
ALT 1 elid 0 -> SoundWire
ALT1 elid C0 -> SSP
ALT1 elid C1 -> DMIC

In other words we cannot test elid in isolation.

ujfalusi replied:
@plbossart, the `ALT 1` was clear, but the `ALT 0` was not.
How a call would look like when looking for HDaudio?
`find_ext2_link(bus, false, 0);` ? (there is no define for HDaudio "elid", there is no such a thing, I know)

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -248,6 +309,113 @@ void hda_bus_ml_free(struct hdac_bus *bus)
+	if (!h2link)
+		return -ENODEV;
+
+	if (sublink >= h2link->slcount)
```
if you have called this with alt==0, then h2link->slcount is 0, no?
This is a extended link only function, the `alt` parameter is useless and it must be 1 in order to have a chance to not return -ENODEV?

plbossart replied:
yes, this is wrong, we do need to initialize the slcount otherwise this is broken.

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -248,6 +309,113 @@ void hda_bus_ml_free(struct hdac_bus *bus)
+		return -ENODEV;
+
+	if (sublink >= h2link->slcount)
+		return -EINVAL;
```
Same here, this will only proceed on extended links?

plbossart replied:
yes, this needs to be initialized

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -535,6 +542,32 @@ int hdac_bus_eml_sdw_sync_go_unlocked(struct hdac_bus *bus)
+	hlink = &h2link->hext_link;
+
+	cmdsync_mask = GENMASK(AZX_REG_ML_LSYNC_CMDSYNC_SHIFT + h2link->slcount - 1,
+			       AZX_REG_ML_LSYNC_CMDSYNC_SHIFT);
```
is there a point of having the `alt` as parameter? h2link->slcount will be 0 if alt is not 1?

plbossart replied:
fixed by initializing, and adding another check on lss

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -228,6 +228,26 @@ static int hdaml_link_shutdown(u32 __iomem *lctl, int sublink)
+		val |= AZX_ML_LCTL_INTEN;
+	else
+		val &= ~AZX_ML_LCTL_INTEN;
+	writel(val, lctl);
```
leave a blank line before as you will do in `hdaml_lctl_offload_enable()`

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -773,6 +773,18 @@ int hda_bus_ml_suspend(struct hdac_bus *bus)
+	if (!h2link)
+		return NULL;
+
+	return &h2link->eml_lock;
```
nitpick: I would be using a positive expression ;)
```c
if (h2link)
	return &h2link->eml_lock;

return NULL;
```

But probably this aligns more with the functions where you are bailing out in case the h2link is not found.

plbossart replied:
yes, I followed the pattern, no change here

ujfalusi reviewed (commented):
@plbossart, the changes are logical and (to me) surprisingly clean and straight forward looking thanks for the great details and thought went into the series.

The only area which is not clear to me is the passing around the `alt` flag to functions which will just fail in case when `alt == false` or in that case will try to use variables which only setup for extended links.

plbossart commented:
@ujfalusi I removed the checks on if (!alt), I am not sure why I did all this in the past. The point is that with alt and elid, you can control power management, interrupt, offload and sync.

See the diffs from v2 here: 
[v3.txt](https://github.com/thesofproject/linux/files/10726300/v3.txt)

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -170,6 +170,69 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
+
+	} while (--retry);
+
+	return -EIO;
```
There is a macro for this which might be good enough:
```c
u32 mask = BIT(sublink) << AZX_ML_LCTL_CPA_SHIFT;;
u32 val;

if (enable)
	return read_poll_timeout(readl, val, val & mask, HDAML_POLL_DELAY_SLACK_US,
				 HDAML_POLL_DELAY_SLACK_US * HDAML_POLL_DELAY_RETRY,
				 true, lctl);
else
	return read_poll_timeout(readl, val, !(val & mask), HDAML_POLL_DELAY_SLACK_US,
				 HDAML_POLL_DELAY_SLACK_US * HDAML_POLL_DELAY_RETRY,
				 true, lctl);
```
it is returning -ETIMEDOUT and not -EIO, but that should be OK.
and subtle differences like HDAML_POLL_DELAY_MIN_US is (HDAML_POLL_DELAY_SLACK_US >> 2) + 1, the timeout is in us and not counts, so it is more precise on that side.

plbossart replied:
humm, not sure. This part is supposed to be relatively simple so that the code can be ported easily to other platforms/environments. Indeed it's open-coding but it's also not supposed to be too Linux-specific. I wasn't able to figure out what the delay was here, no idea what (HDAML_POLL_DELAY_SLACK_US >> 2) + 1 comes from.

@ranj063 what would you prefer?

ujfalusi replied:
@plbossart, it comes from:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/iopoll.h#n55

plbossart replied:
@ujfalusi @ranj063 I will lean on the side of keeping the open-coding for now.

The main reason is that the sequences are similar to existing kernel code in sound/hda/ext and drivers/soundwire, but with the timing adjusted per hardware recommendations.

I think it's better in a first step to keep the code similar to previous functions, and in a second step, if desired, to use an optimized routine.

plbossart replied:
no change.

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -170,6 +170,69 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
+
+	val = readl(lctl);
+	mask = BIT(sublink) << AZX_ML_LCTL_SPA_SHIFT;
+	val |= mask;
```
nitpick: initialize the mask when introduced?

plbossart replied:
yes

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -14,28 +14,239 @@
+		dev_dbg(dev, "Link %d: HDAudio - lsdiid=%d\n",
+			link_idx, hlink->lsdiid);
+
+		h2link->slcount = 1;
```
nitpick: I would have done this before the dev_dbg() print

plbossart replied:
yes

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -251,6 +314,107 @@ void hda_bus_ml_free(struct hdac_bus *bus)
+	if (!h2link)
+		return -ENODEV;
+
+	if (sublink >= h2link->slcount)
```
What is the `sublink` definition for HDaudio? I think the check is about this:
```c
/* ALT = 0 is HDaudio, slcount is not applicable */
if (h2link->alt) {
	if (sublink >= h2link->slcount)
		return -EINVAL;
}
```
Ah, I see, sublink must be 0 for HDaudio otherwise `hdaml_link_init()` would set something else than BIT(16), nevermind, the existing code covers this with the slcount  set to 1.

plbossart replied:
yes, there's no definition for sublink, lscount  was a reserved field read as zero.

ujfalusi commented on include/sound/hda-mlink.h:
```diff
@@ -55,6 +58,12 @@ hdac_bus_eml_power_down_unlocked(struct hdac_bus *bus, bool alt, int elid, int s
+hdac_bus_eml_sdw_power_up_unlocked(struct hdac_bus *bus, int sublink) { return 0; }
+
+static inline int
+hdac_bus_eml_sdw_power_down_unlocked(struct hdac_bus *bus, int sublink) {return 0; }
```
nitpick: missing space before the `return 0`

plbossart replied:
yes

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-mlink.c:
```diff
@@ -255,6 +255,46 @@ static bool hdaml_link_check_interrupt(u32 __iomem *lctl)
+			     HDAML_POLL_DELAY_MIN_US + HDAML_POLL_DELAY_SLACK_US);
+	} while (timeout != 0);
+
+	return -EAGAIN;
```
I think this is an open coded `readl_poll_timeout()` (no need for the `read_poll_timeout()` as you don't sleep before the first read).

plbossart replied:
it's the same argument: simple code that anyone can understand v. optimized macro that requires specific knowledge.

ujfalusi replied:
right, sure. However readl_poll_timeout() is not that complicated to grasp and to be honest, not much need to know the implementation details. It does what it is supposed to.

The only argument to not use it is that we specifically want to open implement and not reuse code already done, which is a valid one, I guess.

plbossart replied:
no change for now, I added an explanation in the commit message

ujfalusi reviewed (commented):
@plbossart, easy and clear to read, I have only few nitpicks and a suggestion to use a generic poll function instead of open coding.

plbossart commented:
Updated version. diff from v3 : 
[1.diff.txt](https://github.com/thesofproject/linux/files/10736366/1.diff.txt)

ujfalusi reviewed (approved):
@plbossart, while I would have used the `readl_poll_timeout()` / `read_poll_timeout()` or our very own `snd_sof_dsp_read_poll_timeout()` instead of open coding, this is not a blocker in any ways.

A pleasure to read PR!

ranj063 commented:
@plbossart why are the CI tests such a disaster?

plbossart commented:
I must have done one simplification too-many @ranj063 @ujfalusi, the nocodec mode is broken and so is the sof-logger tests.

I'll check what's going on.

plbossart commented:
Two errors
a) one is a sof-test issue
rmmod: ERROR: Module snd_hda_ext_core is in use by: snd_sof_intel_hda_mlink
Need to update the code

b) nocodec cannot use hda-mlink, I'll add a quick fix for now.

plbossart commented:
v4: fixed nocodec issue (wrong Kconfig). sof-test will still fail on sof-logger test until https://github.com/thesofproject/sof-test/pull/1003 is merged.

plbossart commented:
The only issue in the tests is the hda-mlink module issue
https://sof-ci.01.org/linuxpr/PR4158/build3217/devicetest/index.html
https://sof-ci.01.org/linuxpr/PR4158/build3218/devicetest/index.html

sof-test was updated, will re-run the tests.

plbossart commented:
SOFCI TEST

plbossart commented:
I don't understand the CI failures, I tested manually on the very same device that fails https://sof-ci.01.org/linuxpr/PR4158/build3223/devicetest/index.html?model=ADLP_RVP_SDW_IPC4ZPH&testcase=check-kmod-load-unload and it's working fine.

Probably some caching issue?

plbossart commented:
SOFCI_TEST

ranj063 commented:
SOFCI TEST

RanderWang reviewed (approved):
LGTM

ujfalusi reviewed (approved):
With the update for sof-test merged, all is good.

plbossart commented:
@bardliao @ranj063 can I bother you to provide your review one final time?

plbossart commented:
GitHub I hate you, why can't I rebase this branch?

"This branch cannot be rebased due to too many changes"

What a useless tool.

plbossart commented:
I'll use a merge commit for now, I don't have time to deal with stupid tools today.

bardliao reviewed (commented):
I am late, but it looks good to me.
